### PR TITLE
Mp telemetry 1.0

### PR DESCRIPTION
**Pull request by Yas-Franco**
Added OpenTelemetry implementation

### CODE REVIEW

**Comment by Yas-Franco** [comment]
#build

**Comment by yasmin-aumeeruddy** [comment]
#push

**Comment by yasmin-aumeeruddy** [comment]
#build

**Comment by yasmin-aumeeruddy** [comment]
#build

**Comment by yasmin-aumeeruddy** [comment]
#build

**Comment by yasmin-aumeeruddy** [comment]
#build

**Comment by Yas-Franco** [comment]
#build

**Comment by Yas-Franco** [comment]
#build

**Comment by Yas-Franco** [comment]
#build

**Comment by Yas-Franco** [comment]
#build

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (fmhwong) [inline .vscode/settings.json:1]: Please add `.vscode` to `.gitignore`

File: dev/cnf/dependabot/check_this_in_if_it_changes/pom.xml [hunk 2]
```diff
@@ -651,6 +801,11 @@
       <artifactId>smallrye-graphql</artifactId>
       <version>1.0.26</version>
     </dependency>
+    <dependency>
```
Comment (fmhwong) [inline dev/cnf/dependabot/check_this_in_if_it_changes/pom.xml]: Do we need this `zipkin` dependency if we use `io.opentelemetry:opentelemetry-exporter-zipkin`?

File: dev/cnf/oss_dependencies.maven [hunk 3]
```diff
@@ -126,6 +156,7 @@ io.smallrye:smallrye-graphql-schema-builder:1.0.26
 io.smallrye:smallrye-graphql-schema-model:1.0.26
 io.smallrye:smallrye-graphql-servlet:1.0.26
 io.smallrye:smallrye-graphql:1.0.26
+io.zipkin.zipkin2:zipkin:2.23.16
```
Comment (fmhwong) [inline dev/cnf/oss_dependencies.maven]: is it a must to ship `zipkin`?

File: dev/com.ibm.websphere.appserver.features/visibility/private/io.openliberty.org.eclipse.microprofile.telemetry-1.0.feature [hunk 4]
```diff
@@ -0,0 +1,9 @@
+-include= ~${workspace}/cnf/resources/bnd/feature.props
+symbolicName=io.openliberty.org.eclipse.microprofile.telemetry-1.0
+visibility=private
+singleton=true
+-features=io.openliberty.mpCompatible-5.0
+-bundles=io.openliberty.org.eclipse.microprofile.telemetry.1.0; location:="dev/api/stable/,lib/"
+kind=ga
```
Comment (fmhwong) [inline dev/com.ibm.websphere.appserver.features/visibility/private/io.openliberty.org.eclipse.microprofile.telemetry-1.0.feature]: We should check this in initially as `noship` first and change it to `beta` after all tests passed.

File: dev/io.openliberty.com.fasterxml.jackson.jr/bnd.bnd [hunk 5]
```diff
@@ -0,0 +1,22 @@
+#*******************************************************************************
+# Copyright (c) 2020 IBM Corporation and others.
```
Comment (fmhwong) [inline dev/io.openliberty.com.fasterxml.jackson.jr/bnd.bnd]: 2020 -> 2022

File: dev/io.openliberty.grpc.1.0.internal.shaded/bnd.bnd [hunk 6]
```diff
@@ -0,0 +1,79 @@
+#*******************************************************************************
+# Copyright (c) 2020, 2022 IBM Corporation and others.
```
Comment (fmhwong) [inline dev/io.openliberty.grpc.1.0.internal.shaded/bnd.bnd]: `2020, 2022` -> `2022` as this is the first check in

File: dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/cdi/OpenTelemetryProducer.java [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+            OpenTelemetrySdk.builder().setTracerProvider(tracerProvider).build();
+
+        Runtime.getRuntime().addShutdownHook(new Thread(tracerProvider::close));
+
+        return openTelemetry;
+    }
+
+    @ApplicationScoped
+    private Resource getServiceName(Map<String,String> oTelConfigs){
+        String serviceName = "open-liberty-service";
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/cdi/OpenTelemetryProducer.java:94]: Should the service name be the app name when neither `otel.service.name` nor `OTEL_SERVICE_NAME` is set?

File: dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/cdi/OpenTelemetryProducer.java [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+        if(oTelConfigs.get("otel.service.name") != null || oTelConfigs.get("OTEL_SERVICE_NAME") != null){
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/cdi/OpenTelemetryProducer.java:96]: Please create String constants for String literals.

File: dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/cdi/OpenTelemetryProducer.java [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+        return Resource.create(Attributes.of(ResourceAttributes.SERVICE_NAME, serviceName));
+    }
+    @ApplicationScoped
+    private SpanExporter getSpanExporter(Map<String,String> oTelConfigs) {
+
+        //default endpoint
+        String endpoint = "http://localhost:14250";
+        String timeout = "10000";
+        if(oTelConfigs.get("otel.traces.exporter") != null){
+            if(oTelConfigs.get("otel.traces.exporter").equals("jaeger") || oTelConfigs.get("OTEL_TRACES_EXPORTER").equals("jaeger")){
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/cdi/OpenTelemetryProducer.java]: Seems like we support "jaeger" only?  Do we need a else if for "zipkin"?

File: dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/cdi/OpenTelemetryProducer.java [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+                if(oTelConfigs.get("otel.exporter.jaeger.endpoint") != null){
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/cdi/OpenTelemetryProducer.java]: This is the wrong order.  Environment variables should have higher priority than system properties according to https://opentelemetry.io/docs/instrumentation/java/automatic/agent-config/#configuring-the-agent

File: dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/cdi/package-info.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+/**
+ * @version 2.0
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/cdi/package-info.java]: 1.0?

File: dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/rest/package-info.java [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+/**
+ * @version 2.0
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/rest/package-info.java]: 1.0?

File: dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/bnd.bnd [hunk 13]
```diff
@@ -0,0 +1,30 @@
+#*******************************************************************************
+# Copyright (c) 2020, 2021 IBM Corporation and others.
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/bnd.bnd]: 2022

File: dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/build.gradle [hunk 14]
```diff
@@ -0,0 +1,14 @@
+/*******************************************************************************
+ * Copyright (c) 2020, 2021 IBM Corporation and others.
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/build.gradle]: 2022

File: dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/config/internal_fat/FATSuite.java [hunk 15]
```diff
@@ -0,0 +1,26 @@
+/*******************************************************************************
+ * Copyright (c) 2020, 2021 IBM Corporation and others.
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/config/internal_fat/FATSuite.java]: 2022

File: dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/config/internal_fat/HotAddMPConfig30.java [hunk 16]
```diff
@@ -0,0 +1,92 @@
+/*******************************************************************************
+ * Copyright (c) 2019, 2021 IBM Corporation and others.
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/config/internal_fat/HotAddMPConfig30.java]: 2022

File: dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/config/internal_fat/apps/hotadd/HotAddMPConfig30Servlet.java [hunk 17]
```diff
@@ -0,0 +1,32 @@
+/*******************************************************************************
+ * Copyright (c) 2019, 2021 IBM Corporation and others.
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/config/internal_fat/apps/hotadd/HotAddMPConfig30Servlet.java]: 2022

File: dev/io.openliberty.microprofile.telemetry.internal.common/.project [hunk 18]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>io.openliberty.microprofile.health.internal.common</name>
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry.internal.common/.project]: Wrong project name?

File: dev/io.openliberty.microprofile.telemetry.internal.common/bnd.bnd [hunk 19]
```diff
@@ -0,0 +1,41 @@
+#*******************************************************************************
+# Copyright (c) 2021 IBM Corporation and others.
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry.internal.common/bnd.bnd]: 2022

File: dev/io.openliberty.microprofile.telemetry.internal.common/bnd.bnd [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+# Contributors:
+#     IBM Corporation - initial API and implementation
+#*******************************************************************************
+-include= ~../cnf/resources/bnd/bundle.props,~../cnf/resources/bnd/transform.props
+
+bVersion=1.0
+
+Bundle-Name: io.openliberty.microprofile.telemetry.internal.common
+Bundle-SymbolicName: io.openliberty.microprofile.telemetry.internal.common
+Bundle-Description: MicroProfile telemetry common Implementation, version ${bVersion}
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry.internal.common/bnd.bnd]: "MicroProfile telemetry common Implementation" -> "MicroProfile Telemetry common implementation"

File: dev/io.openliberty.microprofile.telemetry_git_fat_tck/.project [hunk 21]
```diff
@@ -0,0 +1,40 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>com.ibm.ws.microprofile.config_git_fat_tck</name>
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry_git_fat_tck/.project]: wrong project name

File: dev/io.openliberty.microprofile.telemetry_git_fat_tck/bnd.bnd [hunk 22]
```diff
@@ -0,0 +1,37 @@
+#*******************************************************************************
+# Copyright (c) 2018, 2022 IBM Corporation and others.
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry_git_fat_tck/bnd.bnd]: 2022

File: dev/io.openliberty.microprofile.telemetry_git_fat_tck/bnd.bnd [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+
+src: \
+	fat/src
+
+fat.project: true
+
+// This list does not need to be exhaustive, here we cover the fact that
+// we have   <feature>mpFaultTolerance-${server.xml.feature.version}</feature>
+// in server.xml and also include some currently non existing levels.
+tested.features: mpConfig-1.0,\
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry_git_fat_tck/bnd.bnd]: Should this be mpTelemetry-1.0?

File: dev/io.openliberty.microprofile.telemetry_git_fat_tck/fat/src/com/ibm/ws/microprofile/telemetry/tck/FATSuite.java [hunk 24]
```diff
@@ -0,0 +1,19 @@
+/*******************************************************************************
+ * Copyright (c) 2018 IBM Corporation and others.
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry_git_fat_tck/fat/src/com/ibm/ws/microprofile/telemetry/tck/FATSuite.java]: 2022

File: dev/io.openliberty.microprofile.telemetry_git_fat_tck/fat/src/com/ibm/ws/microprofile/telemetry/tck/TelemetryGitTckLauncher.java [hunk 25]
```diff
@@ -0,0 +1,139 @@
+/*******************************************************************************
+ * Copyright (c) 2018, 2022 IBM Corporation and others.
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry_git_fat_tck/fat/src/com/ibm/ws/microprofile/telemetry/tck/TelemetryGitTckLauncher.java]: 2022

File: dev/io.openliberty.microprofile.telemetry_git_fat_tck/publish/servers/TelemetryGitTCKServer/bootstrap.properties [hunk 26]
```diff
@@ -0,0 +1,13 @@
+###############################################################################
+# Copyright (c) 2018 IBM Corporation and others.
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry_git_fat_tck/publish/servers/TelemetryGitTCKServer/bootstrap.properties]: 2022

File: dev/io.openliberty.microprofile.telemetry_git_fat_tck/publish/servers/TelemetryGitTCKServer/bootstrap.properties [hunk 27]
```diff
@@ -0,0 +1,13 @@
...
+# All rights reserved. This program and the accompanying materials
+# are made available under the terms of the Eclipse Public License v1.0
+# which accompanies this distribution, and is available at
+# http://www.eclipse.org/legal/epl-v10.html
+#
+# Contributors:
+#     IBM Corporation - initial API and implementation
+###############################################################################
+bootstrap.include=../testports.properties
+com.ibm.ws.logging.trace.specification=*=info:APPCONFIG=all
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry_git_fat_tck/publish/servers/TelemetryGitTCKServer/bootstrap.properties:12]: Update the trace spec

File: dev/io.openliberty.microprofile.telemetry_git_fat_tck/publish/servers/TelemetryGitTCKServer/server.xml [hunk 28]
```diff
@@ -0,0 +1,32 @@
+<!--
+    Copyright (c) 2018 IBM Corporation and others.
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry_git_fat_tck/publish/servers/TelemetryGitTCKServer/server.xml]: 2022

File: dev/io.openliberty.microprofile.telemetry_git_fat_tck/publish/tckRunner/tck/src/test/java/com/ibm/ws/microprofile/config/test/ArquillianLoadableExtension.java [hunk 29]
```diff
@@ -0,0 +1,24 @@
+/*******************************************************************************
+ * Copyright (c) 2018 IBM Corporation and others.
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile.telemetry_git_fat_tck/publish/tckRunner/tck/src/test/java/com/ibm/ws/microprofile/config/test/ArquillianLoadableExtension.java]: 2022

File: dev/io.openliberty.microprofile5.internal_fat/bnd.bnd [hunk 30]
```diff
@@ -22,6 +22,6 @@ tested.features: microProfile-5.0
 -buildpath: \
 	io.openliberty.jakarta.servlet.5.0;version=latest,\
 	io.openliberty.jakarta.restfulWS.3.0;version=latest,\
-        io.openliberty.jakarta.annotation.2.0;version=latest,\
+    io.openliberty.jakarta.annotation.2.0;version=latest,\
```
Comment (fmhwong) [inline dev/io.openliberty.microprofile5.internal_fat/bnd.bnd]: Update copyright year

**Comment by Yas-Franco** [comment]
#build

File: dev/io.openliberty.com.fasterxml.jackson.jr/bnd.bnd [hunk 31]
```diff
@@ -0,0 +1,22 @@
...
+
+Bundle-Name: FasterXML Jackson Jr Objects
+Bundle-SymbolicName: io.openliberty.com.fasterxml.jackson.jr
+Bundle-Description: Provides Jackson Jr Objects as a bundle
+
+-includeresource: \
+  @${repo;com.fasterxml.jackson.jr:jackson-jr-objects:jar;2.11.2;EXACT}
+  
+Export-Package: \
+  com.fasterxml.jackson.*;version=2.11.2
```
Comment (tevans78) [inline dev/io.openliberty.com.fasterxml.jackson.jr/bnd.bnd]:
```suggestion
  com.fasterxml.jackson.jr.*;version=2.11.2
```

File: dev/io.openliberty.grpc.1.0.internal.shaded/original.bnd [hunk 32]
```diff
@@ -0,0 +1,15 @@
...
+# are made available under the terms of the Eclipse Public License v1.0
+# which accompanies this distribution, and is available at
+# http://www.eclipse.org/legal/epl-v10.html
+#
+# Contributors:
+#     IBM Corporation - initial API and implementation
+#*******************************************************************************
+
+Bundle-Name: gRPC Shaded
+Bundle-SymbolicName: io.openliberty.grpc.1.0.internal.shaded
```
Comment (tevans78) [inline dev/io.openliberty.grpc.1.0.internal.shaded/original.bnd]:
```suggestion
Bundle-SymbolicName: io.openliberty.io.grpc.netty.shaded
```

File: dev/io.openliberty.grpc.1.0.internal.shaded/transformed.bnd [hunk 33]
```diff
@@ -0,0 +1,15 @@
...
+# which accompanies this distribution, and is available at
+# http://www.eclipse.org/legal/epl-v10.html
+#
+# Contributors:
+#     IBM Corporation - initial API and implementation
+#*******************************************************************************
+-include= ~../cnf/resources/bnd/transform.props
+
+Bundle-Name: gRPC Shaded Jakarta
+Bundle-SymbolicName: io.openliberty.grpc.1.0.internal.shaded.jakarta
```
Comment (tevans78) [inline dev/io.openliberty.grpc.1.0.internal.shaded/transformed.bnd]:
```suggestion
Bundle-SymbolicName: io.openliberty.io.grpc.netty.shaded.jakarta
```

File: dev/io.openliberty.grpc.1.0.internal.shaded/bnd.bnd [hunk 34]
```diff
@@ -0,0 +1,79 @@
...
+  !org.openjsse.javax.net.ssl,\
+  !org.openjsse.net.ssl,\
+  *
+
+Export-Package: \
+  io.grpc.netty.*;version=${grpcVersion},\
+  io.grpc.okhttp.*;version=1.47.0,\
+  okio;version="1.0.0",\
+  okhttp3.*;version="1.0.0"
+  okhttp3;version="1.0.0"
```
Comment (tevans78) [inline dev/io.openliberty.grpc.1.0.internal.shaded/bnd.bnd]: I'm not 100% sure that all of these 3rd party libs should be exported from this one bundle?

File: dev/io.openliberty.grpc.1.0.internal.shaded/original.bnd (hunk 32 above)
Comment (tevans78) [reply dev/io.openliberty.grpc.1.0.internal.shaded/original.bnd]: obviously you would need to rename the project as well

File: dev/io.openliberty.grpc.1.0.internal.shaded/bnd.bnd (hunk 34 above)
Comment (tevans78) [reply dev/io.openliberty.grpc.1.0.internal.shaded/bnd.bnd:54]: in fact, I don't think these exports need to be here at all ... this bundle is not depended on by any other bundle. So probably these should be private.
```suggestion
Private-Package: \
  io.grpc.netty.*;version=${grpcVersion},\
  io.grpc.okhttp.*;version=1.47.0,\
  okio;version="1.0.0",\
  okhttp3.*;version="1.0.0"
  okhttp3;version="1.0.0"
```

File: dev/io.openliberty.grpc.1.0.internal.shaded/bnd.bnd [hunk 35]
```diff
@@ -0,0 +1,79 @@
...
+
+instrument.disabled: true
+
+-fixupmessages.missingexport: "Used bundle version * for exported package";is:=ignore
+
+-buildpath: \
+  io.grpc:grpc-netty-shaded;version=${grpcVersion},\
```
Comment (tevans78) [inline dev/io.openliberty.grpc.1.0.internal.shaded/bnd.bnd]: Also, I'm not sure that we really want to bundle the shaded version of grpc-netty at all... we need to think about this.

**Comment by Yas-Franco** [comment]
#build

File: dev/io.openliberty.org.eclipse.microprofile/telemetry.1.0.bnd [hunk 36]
```diff
@@ -0,0 +1,41 @@
...
+# http://www.eclipse.org/legal/epl-v10.html
+#
+# Contributors:
+#     IBM Corporation - initial API and implementation
+#*******************************************************************************
+
+-include ~../cnf/resources/bnd/transform.props
+bVersion=1.0
+
+Bundle-SymbolicName: io.openliberty.org.eclipse.microprofile.telemetry.1.0; singleton:=true
```
Comment (tevans78) [inline dev/io.openliberty.org.eclipse.microprofile/telemetry.1.0.bnd:15]: This isn't really a microprofile library that is being re-bundled so I think it should be in a project & bundle named `io.openliberty.io.opentelemetry`

**Comment by Yas-Franco** [comment]
#build

**Pull request closed without merging**